raft/src/main/java/org/apache/kafka/raft/FollowerState.java

```diff
@@ -202,16 +205,34 @@ public void setFetchingSnapshot(Optional<RawSnapshotWriter> newSnapshot) {
     }
 
     @Override
-    public boolean canGrantVote(ReplicaKey candidateKey, boolean isLogUpToDate) {
+    public boolean canGrantVote(ReplicaKey replicaKey, boolean isLogUpToDate) {
         log.debug(
             "Rejecting vote request from candidate ({}) since we already have a leader {} in epoch {}",
-            candidateKey,
+            replicaKey,
             leaderId,
             epoch
         );
         return false;
     }
 
+    @Override
+    public boolean canGrantPreVote(ReplicaKey replicaKey, boolean isLogUpToDate) {
+        if (hasFetchedFromLeader) {
+            log.debug(
+                "Rejecting PreVote request from replica ({}) since we already have a leader {} in epoch {}",
+                replicaKey,
+                leaderId,
+                epoch
+            );
+            return false;
+        } else if (!isLogUpToDate) {
+            log.debug(
+                "Rejecting PreVote request from replica ({}) since replica epoch/offset is not up to date with us",
+                replicaKey);
+        }
+        return isLogUpToDate;
+    }
```

Comment:
This method is a lot of lines because we are trying to generate a very specific log message. I don't think we need that since debug message are for developers. How about:
```java
        boolean granting = !hasFetchedFromLeader && isLogUpToDate;
        if (!granting) {
            log.debug(
                "Rejecting PreVote request from replica ({}) since leader {}, epoch is {}, isLogUpToDate is {} and hasFetched is {}",
                replicaKey,
                leaderId,
                epoch,
                isLogUpToDate,
                hasFetchedFromLeader
            );
        }

        return granting;
    }
```

Reply:
thanks, trying out something a bit wordier to make it clear, lmk what you think
```
  "Rejecting PreVote request from replica ({}) either because we have already fetched from leader {} " +
  "in epoch {} (hasFetchedFromLeader={}), or because the replica's log is not as up to date with " +
   "ours (isLogUpToDate={})",
```